### PR TITLE
Prioritize the deployment sources

### LINKED ISSUE #2294: Order of steps in deployment export file incorrect
The 'features' step gets exported, but should appear as the first step in the json file.
Using the json file as is for the recipe file, will result in a failure when executing the recipe.

### PR DESCRIPTION
**Pull request by hishamco**
Fix #2294

### CODE REVIEW

**Comment by hishamco** [comment]
> e.g the features deployment source that would need to run 1st by specifying a very low order.

Good catch

> This by overriding an Order property as you did here. In that case we would just need to define which deployment source need a specific order, e.g the feature deployment source with a very low order, and / or to revise some feature dependencies.

I will have a look ..

**Comment by hishamco** [comment]
@jtkech finally the tests are passed

@sebastienros please :shipit: if you don't have any comments

**Comment by hishamco** [comment]
⌚️ 📆

File: src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs [hunk 1]
```diff
@@ -15,7 +16,9 @@ public AllFeaturesDeploymentSource(IModuleService moduleService)
             _moduleService = moduleService;
         }
 
-        public async Task ProcessDeploymentStepAsync(DeploymentStep step, DeploymentPlanResult result)
+        public override int Order => Int32.MinValue;
+
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs]: LGTM

Couple of comments

For this you could define a very low order as @jtkech suggested.
But not so low that a custom `IDeploymentSource` cannot run before this, a user will want to do this one day ;)

Also can you test, I think, but am not sure, that `ContentDefinitionsStep` also needs to run early (before Contents) or `Contents` can fail.

And lastly, somehow this needs to be reflected in the UI, probably just by sorting based on this Order value, is enough for me. 

Other than that looks like it fixes the issue to me

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs]: > But not so low that a custom IDeploymentSource cannot run before this, a user will want to do this one day ;)

😄 ok, I will use `-100`

> Also can you test, I think, but am not sure, that ContentDefinitionsStep also needs to run early (before Contents) or Contents can fail.

Ok

> And lastly, somehow this needs to be reflected in the UI, probably just by sorting based on this Order value, is enough for me.

This is good and bad, but let us hear from other folks

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Features/Deployment/AllFeaturesDeploymentSource.cs]: -1000 ;)

**Comment by deanmarcussen** [comment]
Also related (for more info) https://github.com/OrchardCMS/OrchardCore/issues/4393

**Comment by hishamco** [comment]
Anything prevent use from merge this?

**Comment by hishamco** [comment]
@jtkech anything prevent from merge this?

**Comment by jtkech** [comment]
LGTM but seems that you didn't apply what @deanmarcussen suggested about running `ContentDefinitionsStep` after Features but before Contents.

**Comment by hishamco** [comment]
I will have a look soon

**Comment by hishamco** [comment]
@jtkech there's no `ContentDefinitionsStep`, there's `ContentDeploymentStep` and `AllContentDeploymentStep`, which one should run after `AllFeaturesDeploymentStep`?

**Comment by jtkech** [comment]
`AllFeaturesDeploymentStep` e.g -1000
Then `ContentDefinitionDeploymentStep` e.g -1000 + 100

Then `AllContentDeploymentStep` and `ContentDeploymentStep` with the same order of e.g -100, but for these 2 last ones maybe we could just use the default order of 0.

**Comment by hishamco** [comment]
I was asking for `ContentDefinitionsStep` not for their values 😊

I will update the pr soon

**Comment by jtkech** [comment]
> I was asking for ContentDefinitionsStep not for their values

In fact it is named `ContentDefinitionDeploymentStep`.

**Comment by hishamco** [comment]
Thanks a lot, pushing the updates ..

**Comment by hishamco** [comment]
Please merge if it looks good to you, I need to move to another PR

**Comment by jtkech** [comment]
The build fails because in the last dev a new `OpenIdServerDeploymentSource : IDeploymentSource` has been created. So you need to merge the last dev and then make this new one inherit from `DeploymentSourceBase`.

**Comment by hishamco** [comment]
Strange!! how it pass in Travis, I will rebase and push again

**Comment by hishamco** [comment]
Believe it or not I forgot that my very old OpenId PR has been merged today :smile:

**Comment by hishamco** [comment]
@jtkech I think we forgot to merge this

**Comment by hishamco** [comment]
I will resolve the conflict now

**Comment by hishamco** [comment]
Please merge once the build is pass and before someone else change another deployment file,  it happens to me all the time  ;)

**Comment by hishamco** [comment]
@sebastienros could you please before someone else change any deployment source ;)

**Comment by hishamco** [comment]
@jtkech could you please merge this, it take so long time, I'm afraid that someone change a deployment file as always happen, so the conflict will raise again :)

**Comment by jtkech** [comment]
@hishamco okay i will re-take a look tomorrow

**Comment by jtkech** [comment]
Okay i just merged your other PR #5207.

For this one, i think it's useful to have order properties and to give them a default value. But, as i remember, to totally fix the "issue" we would need to provide an UI to be able to re-organize the default ordering. So for this one i prefer to let @sebastienros decide ;)

**Comment by hishamco** [comment]
Ok, waiting for Seb ...

**Comment by sebastienros** [comment]
As explained in the issue, not all steps need ordering. Specific one need to be first or last. What about "adding" another interface instead of changing the current one, like `IOrderedDeploymentSource`, if the resolved instance implements it, then we can use it to move it around when it's added. Otherwise it's taken in the order it has been added. And the UI should respect this order. Then we can also (maybe not in this PR) provide custom ordering. At least this way we only change the Feature source.

**Comment by hishamco** [comment]
> And the UI should respect this order

I like the `IOrderedDeployementSource` 
but for now and during the Internet issue in my country, I need to know is the UI responsible for changing the order values?

I will add my comments once the Internet is stable

**Comment by agriffard** [comment]
Please fix the conflict

**Comment by hishamco** [comment]
This need some work except resolving the conflicts, but I will try to fix the conflicts in all m PRs

**Comment by hishamco** [comment]
Seems I did something wrong during the rebase, I will try to revert the duplicate commits

**Comment by hishamco** [comment]
@jtkech could you please have a quick look to the latest changes?

**Comment by hishamco** [comment]
@sebastienros can we consider this to ship in RC2?

File: src/OrchardCore/OrchardCore.Deployment.Core/DeploymentSourceSortedList.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Deployment.Core
+{
+    internal class DeploymentSourceSortedList : IEnumerable<IDeploymentSource>
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Deployment.Core/DeploymentSourceSortedList.cs:8]: You don't need this class. The comparer can just cast a source to IOrderedDeploymentSource to order compared to another one. And call IEnumerable.OrderBy(comparer).

**Comment by deanmarcussen** [comment]
Somewhere along the lines this will need to be reflected in the ui

**Comment by hishamco** [comment]
> Somewhere along the lines this will need to be reflected in the ui

Ya, exactly, Seb mentioned that before, but I'm not sure if it goes in the same PR or we could create another PR extend this

I will try to review Seb earlier comment about `DeploymentSourceSortedList` today

File: src/OrchardCore/OrchardCore.Deployment.Core/DeploymentSourceSortedList.cs (hunk 2 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Deployment.Core/DeploymentSourceSortedList.cs:8]: Seb but I don't think there's a straightforward way to compare `IDeploymentSource` & `IOrderedDeploymentSource`. Am I right?

File: src/OrchardCore/OrchardCore.Deployment.Core/Services/DeploymentManager.cs [hunk 3]
```diff
@@ -23,9 +23,11 @@ public class DeploymentManager : IDeploymentManager
 
         public async Task ExecuteDeploymentPlanAsync(DeploymentPlan deploymentPlan, DeploymentPlanResult result)
         {
-            foreach (var step in deploymentPlan.DeploymentSteps)
+            var orderedDeploymentSources = new DeploymentSourceSortedList(_deploymentSources);
+
+            foreach (var source in orderedDeploymentSources)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Deployment.Core/Services/DeploymentManager.cs:28]: So here, by using an `IComparer<IDeploymentSource>` as suggested above, here we would just need to use an `_deploymentSources.OrderBy(DeploymentSourceComparer)`

File: src/OrchardCore/OrchardCore.Deployment.Core/Services/DeploymentManager.cs [hunk 4]
```diff
@@ -23,9 +23,11 @@ public class DeploymentManager : IDeploymentManager
...
             {
-                foreach (var source in _deploymentSources)
+                foreach (var step in deploymentPlan.DeploymentSteps)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Deployment.Core/Services/DeploymentManager.cs:30]: Not sure a 100% but finally i think that these are the deployment steps that need to be ordered, and this loop would need to be again the main loop as before. I will comment it in the main thread

**Comment by jtkech** [comment]
I did a little review with some suggestions

That said, if at the end we want to re-order things through the admin UI, these are the deployment plans, composed of deployment steps, that are editable. Yes, we could think about another UI listing the sources that we could re-order, and some code to use the related document when executing a deployment.

As @sebastienros said maybe in another PR, but deployment sources (retrieved and first ordered through the DI), are there to execute steps, but the same one could be used to execute multiple steps that would still need to be ordered.

**So, maybe these are the deployment steps that need to be ordered**.

This is when editing a deployment plan that we could re-order its steps, **like widgets are pre-ordered in the layer document**. This way we would just need to execute the steps in the order they were persited for a given plan. Maybe still a defaut `Order` to be used before any ordering, could be part of the abstract `DeploymentStep`, and that could be overridden afterwards.

As a side note, as i remember we have the same kind of issue with the **execution order of recipe steps**.

File: src/OrchardCore/OrchardCore.Deployment.Core/DeploymentSourceSortedList.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+        public void Add(IDeploymentSource deploymentSource)
+        {
+            var orderedDeploymentSource = typeof(IOrderedDeploymentSource).IsAssignableFrom(deploymentSource.GetType())
+                ? (IOrderedDeploymentSource)deploymentSource
+                : new OrderedDeploymentSource(deploymentSource);
+
+            _deploymentSources.Add(orderedDeploymentSource);
+
+            _deploymentSources.Sort(new OrderedDeploymentSourceComparer());
```
[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Deployment.Core/DeploymentSourceSortedList.cs:30]: > it doesn't preserve the initial order for items that are equals

So how `.OrderBy()` is fix this, I expect if all sources have same order, the system treat them equally or we can prioritize them based on their registration in FIFO manner

File: src/OrchardCore/OrchardCore.Deployment.Core/OrderedDeploymentSourceComparer.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+using System.Collections.Generic;
+
+namespace OrchardCore.Deployment.Core
+{
+    internal class OrderedDeploymentSourceComparer : IComparer<IOrderedDeploymentSource>
+    {
+        public int Compare(IOrderedDeploymentSource x, IOrderedDeploymentSource y)
+            => x.Order.CompareTo(y.Order);
```
[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Deployment.Core/OrderedDeploymentSourceComparer.cs:8]: May be, but using `IComparer<IOrderedDeploymentSource>` will avoid the above checks that you wrote

File: src/OrchardCore/OrchardCore.Deployment.Core/Services/DeploymentManager.cs (hunk 3 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Deployment.Core/Services/DeploymentManager.cs:28]: Aha, I understand why you want the above checks ;)

**Comment by hishamco** [comment]
@jtkech thanks for your review and detailed information, IMHO the order of `DeployementSource`s may be done through the code, but it could be done through the UI if it's necessary, but we should allow to sort the `DeploymentSteps`s through the UI only as you mentioned before

I think the PR is almost if not all ready ;) regarding  `DeploymentSteps`s ordering it could be in another PR when we add a UI support, so please let me if there anything else I should do here or we can merge and move forward to UI support

Thanks

**Comment by deanmarcussen** [comment]
The more I think about this, and use deployment steps, the more I feel this should be solely ui based, and at the users discretion.

While `Features` likes to be first generally, I already have recipes / plans where I prefer to activate some features, then set some settings, then activate other features, so it is controllable at my discretion.

I would prefer to see a ui based ordering solution, perhaps with a badge for steps which prefer to be early

**Comment by agriffard** [comment]
Closing it. 
I would also like an ordering from the UI.

**Comment by hishamco** [comment]
I' not sure why this closing ;) as discussed above the ordering from UI could go with another PR

**Comment by hishamco** [comment]
So, what the action here to finish this PR and start with the UI part?

**Comment by deanmarcussen** [comment]
We've discussed this a lot.

For me, there is no action on this PR, i.e. we shouldn't force an Order in Code.

We should, and it'd be a nice PR @hishamco simply support ordering in the UI only.

Perhaps with a badge for steps which prefer to be early.

**Comment by hishamco** [comment]
Ya we discussed a lot,

> For me, there is no action on this PR, i.e. we shouldn't force an Order in Code.

If we remove the order from code, that means the PR is not resolve the related issue, I'm still believe there's a value from UI ordering, I'm thinking of a way to override the code ordering when we do a UI order or what you think @deanmarcussen

> Perhaps with a badge for steps which prefer to be early.

Do you mean display the current order on the badge for now?

**Comment by Skrypt** [comment]
Needs refactoring to order on the UI only.

**Comment by hishamco** [comment]
@Skrypt I remembered Seb asked to do UI ordering in another PR

**Pull request closed without merging**